README.md
<h2 align="center">
  <a href="http://camicroscope.org/"><img src="https://avatars2.githubusercontent.com/u/12075069?s=400&v=4" style="background-color:rgba(0,0,0,0);" height=230 alt="camicroscope: a web-based image viewer optimized for large bio-medical image data viewing"></a>
</h2>

caMicroscope is a web-based biomedical image and data viewer, with a strong emphasis on cancer pathology WSI (Whole Slide Imaging).
This guide has sections for different kinds of use of the platform. The [User Guide](#user-guide) covers the basics on how to use caMicroscope viewer. <a href="https://github.com/SBU-BMI/Nanoborb#nanoborb" target="_blank" rel="noopener noreferrer">nanoBorb</a> covers nanoBorb, the version of caMicroscope designed as a standalone application for individual users without a server. [Hosted Setup](#hosted-setup) covers how to set up caMicroscope for multiple users on a server. [Developer Guide](#developer-guide) covers the broad strokes on how to add new functionality to caMicroscope.

![View Slides](docs/img/View.gif)
![Measure Features](docs/img/Measure.gif)
![Annotate Areas of Interest](docs/img/Draw.gif)
![Alternate Annotation Method](docs/img/Paint.gif)
![Automatic Object Detection](docs/img/Segment.gif)
![Test Classification Models](docs/img/Predict.gif)

# User Guide

## Selecting an Image
Depending on what is providing the image metadata, a different login process may be necessary. For public instances, no log in is necessary, and you can proceed to view slides. Use of other tools, such as annotations may or may not require login in this case.
For slim instances, login should be done through a redirect directly. For pathDB instances, login should be done on the login link on the main page.
At this point, select a collection, if applicable, and proceed to open or "view" the image of your choice.

## Viewing an Image
Once an image is open, you can pan around the image by either clicking and dragging (when no conflicting tool, such as the pen, is open), or by moving the red bounding box in the viewport in the bottom right.
Zooming can be accomplished through the scroll wheel, pinch events on a touch screen, by using the zoom slider or its associated buttons, or by clicking on the zoom number and inputting a different number.

## Using Tools
The toolbar is in the top-left of the main content window. Use the toolbar buttons to manipulate the slide. To close any toolbar button, click the same button again or a new button.

| Tool  | Name        | Function  |
| ----- |-------------| -----|
| ![](https://fonts.gstatic.com/s/i/materialicons/apps/v4/24px.svg)      | Annotations | Opens the Annotation panel, where you can select which annotation set to view, name that annotation set, add optional notes about the annotation set, save the annotation set, and reset the panel to its original state. |
| ![](https://fonts.gstatic.com/s/i/materialicons/view_list/v4/24px.svg)      | Layer Manager      | Opens the Layers Manager panel, where you can select which layers to view. |
| ![](https://fonts.gstatic.com/s/i/materialicons/home/v4/24px.svg)      | Home      | Return to the data table so that you can open another slide.|
| ![](https://fonts.gstatic.com/s/i/materialicons/create/v4/24px.svg)      | Draw      |  Draw thin lines, thick lines, or polygons on the image. Annotations can also be computer aided using the Smart-pen tool. Draw them, stretch them, remove them. To maintain the integrity of measurements, avoid drawing shapes that overlap or intersect one another. |
| ![](https://fonts.gstatic.com/s/i/materialicons/colorize/v4/24px.svg)      | Preset Labels      |  Use a preset annotation type immediately to quickly annotate a slide consistently. |
| ![](https://fonts.gstatic.com/s/i/materialicons/search/v4/24px.svg)       | Magnifier      |The Magnifier works like a magnifying glass and allows you to see the slide at normal magnification (1.0), low magnification (0.5), or high magnification (2.0). Click a magnification level and place the bounding box on the area of the slide you want to magnify. |
| ![](https://fonts.gstatic.com/s/i/materialicons/space_bar/v4/24px.svg)      | Measurement      | Drag this tool on the slide to learn the measurement in micrometers. |
| ![](https://fonts.gstatic.com/s/i/materialicons/share/v4/24px.svg)      | Share View      |Opens a window with a URL to the current presentation state of the slide including the magnification level, layers that are currently open, and your position on the image.|
| ![](https://fonts.gstatic.com/s/i/materialicons/view_carousel/v4/24px.svg)      | Side by Side Viewer     |Shows the Layer Manager panel, the left and right layers, and inset window. For the right and left layer, select which layer you want to view. |
| ![](https://fonts.gstatic.com/s/i/materialicons/satellite/v4/24px.svg)      | Heatmap     | For a slide with heatmap data, opens the choices of heatmaps available, as well as ways of displaying the heatmaps. The gradient shows all of the values on the selected spectrum for the field you selected. Contains a heatmap edit pen function.|
| ![](https://fonts.gstatic.com/s/i/materialicons/label/v4/24px.svg)      | Labeling      |Use this tool to draw a circle or rectangle around a tumor region, measure an area on the slide, download labels, and submit a bug report. The Labeling tool has its own toolbar with tools in the following order from left to right: return to the previous slide, place a square on the slide, place a circle on the slide, measure an area, download labels, and submit a bug report. Click the left arrow at the far right of the toolbar to hide it, then click the right arrow to show it. |
| ![](https://fonts.gstatic.com/s/i/materialicons/timeline/v6/24px.svg)      | Segment      | This tool allows you to display, count, and export nuclear segmentations on the image. Clicking this tool opens the following custom toolbar. |
| ![](https://fonts.gstatic.com/s/i/materialicons/aspect_ratio/v4/24px.svg)      | Model      | Show results from a pre-trained tensorflow compatible model on a ROI of the slide. |
| ![](https://fonts.gstatic.com/s/i/materialicons/get_app/v4/24px.svg)      | Download Slide      | Download the slide image to your system |
| ![](https://fonts.gstatic.com/s/i/materialicons/playlist_add_check/v8/24px.svg)      | Mark Reviewed      | Use to signify the completion of review of a slide. |
| ![](https://fonts.gstatic.com/s/i/materialicons/bug_report/v4/24px.svg)      | Bug Report      | Report a bug or give feedback. |
| ![](https://fonts.gstatic.com/s/i/materialicons/camera_enhance/v4/24px.svg)      | Slide Capture      | Click to take a screenshot of the slide and annotations on it. |
| ![](https://fonts.gstatic.com/s/i/materialicons/help/v4/24px.svg)      | Tutorial      | Click to view a guided tour of the viewer tools. |


## Toolbar Shortcuts

| Tool         | Shortcut  |
|------------- |-----------|
| Annotation   |  Ctrl + a |
| Magnifier    |  Ctrl + m |
| Measurement    |  Ctrl + r |
| Side-by-Side |  Ctrl + s |
| Close all tools |  ESC   |

# Hosted Setup
The full distribution repository for hosted caMicroscope is [here](https://github.com/camicroscope/Distro/).
run with `docker-compose -f caMicroscope.yml up`

this will build all services and run in the foreground.
Use `docker-compose -f caMicroscope.yml build` to rebuild the services.

Once everything is up, go to \<the host this is running on\>:4010/ to see the landing page.

# Other Resources
- **Slack:** <http://bit.ly/camicroscope>
- **Discussion mailing list:** <https://groups.google.com/g/camicroscope>
- **Sample Tensorflow Models:** <https://github.com/camicroscope/tfjs-models>

# Developer Guide
We are collecting feedback to write this section in more detail. Please add your suggestions [here](https://github.com/camicroscope/caMicroscope/issues/267).

caMicroscope is open source software. Any involvement and contribution with the caMicroscope project is greatly appreciated. Feel free to get directly involved in any of the repositories in the caMicroscope organization. New developers may find the notes in [CONTRIBUTING](https://github.com/camicroscope/caMicroscope/blob/master/CONTRIBUTING.md) helpful to start contributing to caMicroscope.

It is highly recommended to make any changes off of the develop branch of a repository, and, when ready, create a PR to the develop branch of the source repository. Before sending the PR, make sure that there are no linting errors by running ```npm install``` and then ```npm run lint```  to see the errors and ```npm run lint-fix``` to automatically fix the errors in the repository folder.

Source code organization ie the file structure of caMicroscope can be found in [file structure](https://github.com/camicroscope/caMicroscope/blob/master/docs/file_structure.md)

## Fast Local Changes
When using the hosted setup, you can have the distribution host the changes from your local. Follow these steps :
- Clone this repository, the [Caracal repository](https://github.com/camicroscope/Caracal/) and [the distribution](https://github.com/camicroscope/Distro/) in the same parent directory
- Set the build to build your local changes inste ad of the hosted git versions by editing the ca-back container section of your develop.yml. Replace the build context section with the path to your caracal checkout ("../Caracal"), and add `- ../caMicroscope:/src/camicroscope` to the volumes.
- Remove this line from 'Dockerfile' in Caracal repository :
```
RUN git clone https://github.com/${fork:-camicroscope}/camicroscope.git --branch=${viewer:-master}
```
- In Distro repository, enter the following commands :
```
docker-compose -f develop.yml build
docker-compose -f develop.yml up
```


HISTORY.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### caMicroscope [Unreleased](https://github.com/camicroscope/camicroscope/compare/v3.11.0...camicroscope:develop)
###### TBD

### caMicroscope [3.11.0](https://github.com/camicroscope/camicroscope/compare/v3.10.2...camicroscope:v3.11.0)
###### 2023-11-17
* Dicom and Bioformats Intergration (#647, #653, #655)
* Improved Machine Learning Toolkit (#658)

### caMicroscope [3.10.2](https://github.com/camicroscope/camicroscope/compare/v3.10.1...camicroscope:v3.10.2)
###### 2023-06-16
* point to point
* Fixed Workbench Navbar Overlapping  #580
* The Reload button is not Responsive for small size devices #579
* style: art Fixed fat hamburger icon & alignment issues of dropdown issue#596 #597
* Styled export.html  #601
* fix: bug Responsiveness issues of workbench section #605 #605
* Slight spelling mistake caught while browsing code #630
* Adding Footer to Sign Up Page #635
* Fixed the UI of admin page: Solved issue #621 #620
* Fixed spelling mistake #614
* FIXED Overlapping / Not properly styled on class=card in the workbench.html #616
* Fixed import.html interface #613

### caMicroscope [3.10.1](https://github.com/camicroscope/camicroscope/compare/v3.10.0...camicroscope:v3.10.1)
###### 2022-11-18
* Bugfix for Magnifying glass with slide token auth

### caMicroscope [3.10.0](https://github.com/camicroscope/camicroscope/compare/v3.9.11...camicroscope:v3.10.0)
###### 2022-10-25
* Result Export Page [caMicroscope/566](https://github.com/camicroscope/caMicroscope/pull/566)
* Slide Token injection [caMicroscope/567](https://github.com/camicroscope/caMicroscope/pull/567)


### caMicroscope [3.9.11](https://github.com/camicroscope/camicroscope/compare/v3.9.10...camicroscope:v3.9.11)
###### 2021-01-20
* add redir page/app

### caMicroscope [3.9.10](https://github.com/camicroscope/camicroscope/compare/v3.9.9...camicroscope:v3.9.10)
###### 2021-01-20
* pathdb mode improvements

### caMicroscope [3.9.9](https://github.com/camicroscope/camicroscope/compare/v3.9.8...camicroscope:v3.9.9)
###### 2021-11-08
* Fix and enhance Human Readable URL support

### caMicroscope [3.9.10](https://github.com/camicroscope/camicroscope/compare/v3.9.9...camicroscope:v3.9.10)
###### 2021-01-20
* Human Readable Urls for pathdb

### caMicroscope [3.9.9](https://github.com/camicroscope/camicroscope/compare/v3.9.8...camicroscope:v3.9.9)
###### 2021-11-08
* Fix and enhance Human Readable URL support

### caMicroscope [3.9.8](https://github.com/camicroscope/camicroscope/compare/v3.9.6...camicroscope:v3.9.8)
###### 2021-10-21
* Add Segmentation -> Annotation Functionality [#542](https://github.com/camicroscope/caMicroscope/pull/542)
* Margin and Table UI Fixes [#549](https://github.com/camicroscope/caMicroscope/pull/549) [#551](https://github.com/camicroscope/caMicroscope/pull/551)
* Add Pathdb Quad Lookup
* Add Docs link to landing
* retry slides one time if they have an error

### caMicroscope [3.9.6](https://github.com/camicroscope/camicroscope/compare/v3.9.4...camicroscope:v3.9.6)
###### 2021-08-05
* Some integration with the "Smartpen" concept [#526](https://github.com/camicroscope/caMicroscope/pull/526), [#532](https://github.com/camicroscope/caMicroscope/pull/532)
* Preset label bugfix

### caMicroscope [3.9.4](https://github.com/camicroscope/camicroscope/compare/v3.9.3...camicroscope:v3.9.4)
###### 2021-06-16
* Github action for ci testing [#514](https://github.com/camicroscope/caMicroscope/pull/514)
* Bootstrap Upgrades [#518](https://github.com/camicroscope/caMicroscope/pull/518), [#519](https://github.com/camicroscope/caMicroscope/pull/519)
* Migrate Additional Links to mongo from conf file
* Table Style Fix for large number of pages

### caMicroscope [3.9.3](https://github.com/camicroscope/camicroscope/compare/v3.9.2...camicroscope:v3.9.3)
###### 2021-04-15
* Bugfix for additional links url
* Additional links can launch with slide id

### caMicroscope [3.9.2](https://github.com/camicroscope/camicroscope/compare/v3.9.1...camicroscope:v3.9.2)
###### 2021-04-07
* Github action for ci testing [#502](https://github.com/camicroscope/caMicroscope/pull/502)
* Validate Emails on signup [#501](https://github.com/camicroscope/caMicroscope/pull/501)
* Image Enhancements [#500](https://github.com/camicroscope/caMicroscope/pull/500)
* Configurable Toolbar Links [#511](https://github.com/camicroscope/caMicroscope/pull/511)
* Documentation Improvements [#506](https://github.com/camicroscope/caMicroscope/pull/506)


### caMicroscope [3.9.1](https://github.com/camicroscope/camicroscope/compare/v3.8.4...camicroscope:v3.9.1)
###### 2021-03-05
* Bugfix: Annotation Name Filtering

### caMicroscope [3.9.0](https://github.com/camicroscope/camicroscope/compare/v3.8.4...camicroscope:v3.9.0)
###### 2021-03-04
* Asynchronous Layer Manager
* Auto convert png and jpg files via table load.
* Undo Draw Button [#491](https://github.com/camicroscope/caMicroscope/pull/491)
* Addition of screenshot feature to the viewer [#496](https://github.com/camicroscope/caMicroscope/pull/496)
* Table Button Responsiveness [#477](https://github.com/camicroscope/caMicroscope/pull/477)
* Improved Documentation [#478](https://github.com/camicroscope/caMicroscope/pull/478), [#488](https://github.com/camicroscope/caMicroscope/pull/488)
* Tour, save state and resume [#487](https://github.com/camicroscope/caMicroscope/pull/487)
* Toolbar Dependency Bugfix [#494](https://github.com/camicroscope/caMicroscope/pull/494)

### caMicroscope [3.8.4](https://github.com/camicroscope/camicroscope/compare/v3.8.3...camicroscope:v3.8.4)
###### 2021-01-25
* Addition of a collection type, and related panel in table
* Google Drive Upload Support [#467](https://github.com/camicroscope/caMicroscope/pull/467)

js/init.js
// CAMIC is an instance of camicroscope core
// $CAMIC in there
let $CAMIC = null;
// for all instances of UI components
const $UI = new Map();

const $D = {
  pages:{
    home:'./table.html',
    table:'./table.html'
  },
  params:null, // parameter from url - slide Id and status in it (object).
  overlayers:null, // array for each layers
  templates:null // json schema for prue-form
};


// initialize viewer page
function initialize(){

  // init UI -- some of them need to wait data loader to load data
  // because UI components need data to initialize
  initUIcomponents();

  // create a viewer and set up
  initCore();

  // loading the form template data
  FormTempaltesLoader();

  // loading the overlayers data
  OverlayersLoader();


}

function loadData(){

}

// setting core functionalities
function initCore(){
  // start initial
  // TODO zoom info and mmp
  const opt = {
    draw:{
      // extend context menu btn group
      btns:[
        { // annotation
          type:'btn',
          title:'Annotation',
          class:'material-icons',
          text:'description',
          callback:saveAnnotation
        },
        { // analytics
          type:'btn',
          title:'Analytics',
          class:'material-icons',
          text:'settings_backup_restore',
          callback:saveAnalytics
        }
      ]
    }
  }
  // set states if exist
  if($D.params.states){
    opt.states = $D.params.states;
  }

  try{
    let slideQuery = {}
    slideQuery.id = $D.params.slideId
    slideQuery.name = $D.params.slide
    slideQuery.location = $D.params.location
    $CAMIC = new CaMic("main_viewer", slideQuery, opt);
  }catch(error){
    Loading.close();
    $UI.message.addError('Core Initialization Failed');
    console.error(error);
    return;
  }

  $CAMIC.loadImg(function(e){
    // image loaded
    if(e.hasError){
      $UI.message.addError(e.message)
    }else{
      $D.params.data = e;
      // popup panel
      $CAMIC.viewer.addHandler('canvas-lay-click',function(e){
        if(!e.data) {
          $UI.annotPopup.close();
          return;
        }
        // for support QUIP 2.0
        const data = Array.isArray(e.data)? e.data[e.data.selected]: e.data;

        const type = data.provenance.analysis.source;
        let body;
        let attributes;
        switch (type) {
          case "human":
            // human
            attributes = data.properties.annotations;
            body = convertHumanAnnotationToPopupBody(attributes);
            $UI.annotPopup.showFooter();
            break;
          case "computer":
            // handle data.provenance.analysis.computation = `segmentation`
            attributes = data.properties.scalar_features[0].nv;
            body = {type:'map',data:attributes};
            $UI.annotPopup.hideFooter();
            break;
          default:
            return;
            // statements_def
            break;
        }
        $UI.annotPopup.data = {
          id:data.provenance.analysis.execution_id,
          oid:data._id.$oid,
          annotation:attributes
        };
        $UI.annotPopup.setTitle(`id:${data.provenance.analysis.execution_id}`);
        $UI.annotPopup.setBody(body);
        $UI.annotPopup.open(e.position);
      });

      // create the message bar TODO for reading slide Info TODO
      $UI.slideInfos = new CaMessage({
      /* opts that need to think of*/
        id:'cames',
        defaultText:`Slide: ${$D.params.data.name}`
      });

      // spyglass
      $UI.spyglass = new Spyglass({
        targetViewer:$CAMIC.viewer,
        imgsrc:$D.params.data.url
      });
    }
  });

  $CAMIC.viewer.addHandler('open',function(){
    $CAMIC.viewer.canvasDrawInstance.addHandler('start-drawing',startDrawing);
    $CAMIC.viewer.canvasDrawInstance.addHandler('stop-drawing',stopDrawing);
  });
}


// initialize all UI components
function initUIcomponents(){
  /* create UI components */


  // create the message queue
  $UI.message = new MessageQueue();

  // create the tool bar
  $UI.toolbar = new CaToolbar({
  /* opts that need to think of*/
    id:'ca_tools',
    zIndex:601,
    mainToolsCallback:mainMenuChange,
    subTools:[
      // home
      {
        icon:'home',// material icons' name
        title:'Home',
        type:'btn',// btn/check/dropdown
        value:'home',
        callback:goHome
      }, //
      // pen
      {
        icon:'create',// material icons' name
        title:'Draw',
        type:'multistates',
        callback:draw
      },
      // magnifier
      {
        icon:'search',
        title:'Magnifier',
        type:'dropdown',
        value:'magn',
        dropdownList:[
          // free draw
          {
            //icon:'linear_scale',
            value:10,
            title:'10x',
            checked:true
          },
          // rectangle fraw
          {
            //icon:'timeline',
            value:20,
            title:'20x'
          },
          {
            //icon:'timeline',
            value:40,
            title:'40x'
          }
        ],
        callback:toggleMagnifier
      },
      // measurement tool
      {
        icon:'space_bar',
        title:'Measurement',
        type:'check',
        value:'measure',
        // dropdownList:[
        //   // free draw
        //   {
        //     icon:'linear_scale',
        //     value:'straight',
        //     title:'Line',
        //     checked:true
        //   },
        //   // rectangle draw
        //   {
        //     icon:'timeline',
        //     value:'coordinate',
        //     title:'Coordinate'
        //   }
        // ],
        callback:toggleMeasurement
      },
      // download TODO
      // {
      //   icon:'file_download',
      //   title:'Download Image',
      //   type:'btn',
      //   value:'download',
      //   callback:imageDownload
      // },
      // share
      {
        icon:'share',
        title:'Share View',
        type:'btn',
        value:'share',
        callback:shareURL
      },
      {
        icon:'view_carousel',
        title:'Side By Side Viewer',
        value:'dbviewers',
        type:'check',
        callback:toggleViewerMode
      }

    ]
  });

  // create two side menus for tools
  $UI.appsSideMenu = new SideMenu({
    id:'side_apps',
    width: 300,
    //, isOpen:true
    callback:toggleSideMenu
  });

  $UI.layersSideMenu = new SideMenu({
    id:'side_layers',
    width: 300,
    contentPadding:5,
    //, isOpen:true
    callback:toggleSideMenu
  });

  /* annotation popup */
  $UI.annotPopup = new PopupPanel({
    footer:[
      // { // edit
      //   title:'Edit',
      //   class:'material-icons',
      //   text:'notes',
      //   callback:anno_edit
      // },
      { // delete
        title:'Delete',
        class:'material-icons',
        text:'delete_forever',
        callback:anno_delete
      }
    ]
  });

  var checkOverlaysDataReady = setInterval(function () {
    if($D.overlayers) {
      clearInterval(checkOverlaysDataReady);

      // create UI and set data
      $UI.layersViewer = new LayersViewer({id:'overlayers',data:$D.overlayers,sortChange:sort_change,callback:callback });

      callback($D.overlayers.filter(lay => lay.isShow));

      $UI.layersViewer.elt.parentNode.removeChild($UI.layersViewer.elt);

      // add to layers side menu
      const title = document.createElement('div');
      title.classList.add('item_head');
      title.textContent = 'Layers Manager';
      $UI.layersSideMenu.addContent(title);
      $UI.layersSideMenu.addContent($UI.layersViewer.elt);
    }
  }, 500);






  var checkTemplateSchemasDataReady = setInterval(function () {
    if($D.templates) {
      clearInterval(checkTemplateSchemasDataReady);
      const annotRegex = new RegExp('annotation', 'gi');
      const annotSchemas = $D.templates.filter(item=> item.id.match(annotRegex));
      /* annotation control */
      $UI.annotOptPanel = new OperationPanel({
        //id:
        //element:
        formSchemas:annotSchemas,
        action:{
          title:'Save',
          callback:anno_callback
        }
      });
      $UI.appsList.clearContent('annotation');
      $UI.appsList.addContent('annotation',$UI.annotOptPanel.elt);

      /* algorithm control */
      const algoRegex = new RegExp('algo', 'gi');
      const algoSchemas = $D.templates.filter(item => item.id.match(algoRegex));
      $UI.algOptPanel = new OperationPanel({
        //id:
        //element:
        title:'Algorithm:',
        formSchemas:algoSchemas,
        action:{
          title:'Run',
          callback:algo_callback
        }
      });
      $UI.appsList.clearContent('analytics');
      $UI.appsList.addContent('analytics',$UI.algOptPanel.elt);
      $UI.appsList.addContent('analytics', AnalyticsPanelContent);

    }
  }, 500);



    // collapsible list
  $UI.appsList = new CollapsibleList({
    id:'collapsiblelist',
    list:[
      {
        id:'annotation',
        title:'Annotation',
        icon:'border_color',
        content: "No Template Loaded" //$UI.annotOptPanel.elt
        // isExpand:true

      },{
        id:'analytics',
        icon:'find_replace',
        title:'Analytics',
        content:"No Template Loaded" //$UI.algOptPanel.elt,
      }
    ],
    changeCallBack:getCurrentItem
  });


  // detach collapsible_list
  $UI.appsList.elt.parentNode.removeChild($UI.appsList.elt);
  $UI.appsSideMenu.addContent($UI.appsList.elt);

  $UI.multSelector = new MultSelector({id:'mult_selector'});
  $UI.multSelector.addHandler('cancel',multSelector_cancel);
  $UI.multSelector.addHandler('action',multSelector_action);
}

function redirect(url ,text = '', sec = 5){
  let timer = sec;
  setInterval(function(){
    if(!timer) {
      window.location.href = url;
    }

    if(Loading.instance.parentNode){
      Loading.text.textContent = `${text} ${timer}s.`;
    }else{
      Loading.open(document.body,`${text} ${timer}s.`);
    }
    // Hint Message for clients that page is going to redirect to Flex table in 5s
    timer--;

  }, 1000);
}
